lib/3scale/backend/stats/cleaner.rb

```diff
@@ -106,8 +96,51 @@ def delete!(redis_conns, log_deleted_keys: nil)
             logger.info("Finished deleting the stats keys for these services: #{services.to_a}")
           end
 
+          # Deletes all the stats keys set to 0.
+          #
+          # Stats keys set to 0 are useless and occupy Redis memory
+          # unnecessarily. They were generated due to a bug in previous versions
+          # of Apisonator.
+          # Ref: https://github.com/3scale/apisonator/pull/247
+          #
+          # As the .delete function, this one also receives a collection of
+          # instantiated Redis clients and those need to connect to Redis
+          # servers directly.
+          #
+          # @param [Array] redis_conns Instantiated Redis clients.
+          # @param [IO] log_deleted_keys IO where to write the logs. Defaults to
+          #             nil (logs nothing).
+          def delete_stats_keys_set_to_0(redis_conns, log_deleted_keys: nil)
+            redis_conns.each do |redis_conn|
```

Comment:
We could use something like `partition` to collect successul and unsuccessful connections and output the details (including URIs) of each category. Wdyt?

Reply:
`logger.error` already tells you the details of the connection that failed

Reply:
Good - what about reporting the URIs that failed after the rake task completes?

Reply:
Why wait until the end?

Reply:
To make it easy for the user to know which ones should be retried - in particular if there is a screenful of errors.

Reply:
How can there be a screenful of errors? How many shards do you think users have for this? :scream:

Reply:
I'm not thinking about shards. I'm thinking about backtraces when any one of those `raise exception` lines is executed, say, after yet any other shard failed. Or having a shard fail while logging the deleted keys to stdout, and then logging even more deleted keys, losing the message in the middle of a million screens and having to go hunt for it manually if your terminal still has it in its buffer.

Reply:
If the code gets to the `raise exception` line it doesn't continue with other shards. It stops the execution. It's done that way because if it gets to that line, something unexpected (possible bug?) happened. It's not for things that we might expect like connection errors, timeouts, etc.

There can be at most 1 error message per shard. It could get mixed with the keys as you said, but keep in mind that the keys are written to STDOUT and the errors to STDERR.

Reply:
I'm aware, I am thinking that the usual case of people launching this task will be with both stderr and stdout going to the same terminal output. This is a suggestion that I think will improve the usability and be appreciated by the guy launching the task at a low cost in code changes, but if you think it's not worth it then let's resolve this thread.

Reply:
Wait, this made me realize that what we are doing in `handle_redis_exception` is not correct. When there's an error, we should rescue, retry, log the error if all the attempts failed, and continue with the next shard. Right now the code continues with the same one.

Reply:
Added a commit implementing your suggestion about logging only at the end: https://github.com/3scale/apisonator/pull/250/commits/c836a3dfe5b66d0310bc998c6842b9c040f716f1